loky/backend/context.py

```diff
@@ -113,15 +119,58 @@ def cpu_count():
        set by docker and similar container orchestration systems);
      * the value of the LOKY_MAX_CPU_COUNT environment variable if defined.
     and is given as the minimum of these constraints.
+
+    If ``only_physical_cores`` is True, return the number of physical cores
+    instead of the number of logical cores (hyperthreading / SMT). Note that
+    this option is not enforced if the number of usable cores is controlled in
+    any other way such as: process affinity, restricting CFS scheduler policy
+    or the LOKY_MAX_CPU_COUNT environment variable. If the number of physical
+    cores is not found, return the number of logical cores.
+ 
     It is also always larger or equal to 1.
     """
-    import math
-
+    # TODO: use os.cpu_count when dropping python 2 support
     try:
         cpu_count_mp = mp.cpu_count()
     except NotImplementedError:
         cpu_count_mp = 1
 
+    cpu_count_user = _cpu_count_user(cpu_count_mp)
+    aggregate_cpu_count = min(cpu_count_mp, cpu_count_user)
+
+    if only_physical_cores:
+        cpu_count_physical, error = _count_physical_cores()
+        if cpu_count_user < cpu_count_mp:
+            # Respect user setting
+            cpu_count = max(cpu_count_user, 1)
+        elif cpu_count_physical == "not found":
+            # Fallback to default behavior
+            if error is not None:
+                # warns only the first time
+                warnings.warn(
+                    "Could not find the number of physical cores. Returning "
+                    "the number of logical cores instead. You can silence "
+                    "this warning by setting LOKY_MAX_CPU_COUNT to the number "
+                    "of cores you want to use.\n")
+                if isinstance(error, str):
+                    print(error)
+                elif sys.version_info >= (3, 5):
+                    # TODO remove the version check when dropping py2 support
+                    traceback.print_tb(error.__traceback__)
```

Comment:
I would rather have `_count_physical_cores` return `cpu_count_physical, error_cause, exception` and then integrate `error_cause` as part of the warning message and print the traceback of exception on Python 3.+ independently instead of doing `print(error)` outside of the warning.

Reply:
done, here's how it renders:
```
/home/jeremie/R/O/loky/loky/backend/context.py:150: UserWarning: Could not find the number of physical cores for the following reason:
ceci est un message d'erreur
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  warnings.warn(
  File "/home/jeremie/R/O/loky/loky/backend/context.py", line 226, in _count_physical_cores
    raise AttributeError("ceci est un message d'erreur")
```